### PR TITLE
The big one; settings panel, darkmode, tweaks

### PR DESCRIPTION
**Pull request by NoSloppy**

### CODE REVIEW

File: style_editor.html [hunk 1]
```diff
@@ -274,7 +482,7 @@
 </style>
 
 
-<script id="vertex-shader" type="x-shader/x-vertex">
+<script id="vertex_shader" type="x-shader/x-vertex">
```
Comment (NoSloppy) [inline style_editor.html:485]: Universalizing: ID's use underscores, classes use dashes

File: style_editor.html [hunk 2]
```diff
@@ -287,7 +495,7 @@
 }
 </script>
 
-<script id='hilt_graflex' type='x-shader/x-fragment'>
+<script id="hilt_graflex" type="x-shader/x-fragment">
```
Comment (NoSloppy) [inline style_editor.html:498]: Universalizing: double quotes like previous code had throughout.
Plenty more of this eventually; whitespace, indention , clean up in general.
Nothing wrong with perfection?

File: style_editor.html [hunk 3]
```diff
@@ -9251,17 +9471,6 @@
   blade.addEffect(STATE_ON ? EFFECT_IGNITION : EFFECT_RETRACTION, Math.random() * 0.7 + 0.2);
 }
 
-function UpdateLockupButtons() {
```
Comment (NoSloppy) [inline style_editor.html:9254]: Unused function. Elements are drop down menu options. Maybe they were buttons originally.

File: style_editor.html [hunk 4]
```diff
@@ -9280,37 +9489,199 @@
...
   STATE_LOCKUP = STATE_LOCKUP == LOCKUP_LIGHTNING_BLOCK ? 0 : LOCKUP_LIGHTNING_BLOCK;
-  UpdateLockupButtons();
 }
 
 function ClickMelt() {
   STATE_LOCKUP = STATE_LOCKUP == LOCKUP_MELT ? 0 : LOCKUP_MELT;
-  UpdateLockupButtons();
 }
 
-function ClickInhilt() {
```
Comment (NoSloppy) [inline style_editor.html:9303]: Inhilt and Slow moved into Settings panel.

File: style_editor.html [hunk 5]
```diff
@@ -9679,17 +10190,17 @@
...
+            min="0" max="32768" step="1" value="0"
+            oninput="updateVariantValue(this.value)" >
           <input
             id="VARIANT_MINUS" name=varminus type=button value="<" 
             title="Decreases the Variant value. Hold to increase speed of the changing value."
             onmousedown="startAdjustingValue(-1, 'VARIANT_SLIDER')"
             onmouseup="stopAdjustingValue()"
             onmouseleave="stopAdjustingValue()" >
-          <input
-            id="VARIANT_SLIDER" name=varslider type=range class="variant-slider"
```
Comment (NoSloppy) [inline style_editor.html:9689]: Moved decrease button next to value

File: style_editor.html [hunk 6]
```diff
@@ -9748,18 +10258,69 @@
...
+    <div id=TABS class="tabs-container">
+      <a
+        href="#" id="SETTINGS_BUTTON" class="settings-button"
+        title="Settings"
+        onclick="toggleSettingsPanel()" >
+      </a>
+      <div id="settings_panel" class="settings-panel"
+        title="Turning on checkboxes saves your settings for future visits to the Style Editor page." >
+
+        <div class="settings-section">
```
Comment (NoSloppy) [inline style_editor.html:10273]: Added categories. General and Blade Preview sections...for now.

File: style_editor.html [hunk 7]
```diff
@@ -9280,37 +9490,220 @@
...
+  // WAVLEN_VALUE
+  var wavlenInput = FIND("WAVLEN_VALUE");
+  wavlenValue = parseInt(getSavedCookie("wavlenValue"));
+  if (!wavlenValue) {
+    wavlenValue = 500;
+  }
+  // console.log("BC-getSavedCookie called, wavlenValue = " + wavlenValue);
+  wavlenInput.value = wavlenValue;
+  myWavLen.setLength(wavlenValue);
+wavlenInput.addEventListener("click", function(e) {
```
Comment (NoSloppy) [inline style_editor.html:9635]: This may look like a lot of listeners, but they make the value entry better.
Single clicking in the field selects all digits, so just starting typing starts a total new entry as opposed to requiring clicking an insertion point and needing to backspace or delete.
The input value is safeguarded against leaving an empty value, which can cause issues.
If all digits deleted, a minimum value of 1 is autofilled, and left selected so that entry of a different value is easier.
Hitting enter "enters" the value, just removing focus from the box.

File: style_editor.html [hunk 8]
```diff
@@ -9718,14 +10254,13 @@
...
 
-      <br>
         <pre><textarea 
           rows=10 cols=80 id=style wrap="off" class="textbox"
           title="Style and value editing area. Edits in here are committed by clicking the Submit button below." >
         </textarea></pre>
         <span class="middle-btn-container">
           <input
-            type=button name=run value="Submit"
+            type=button name=run value="Submit" class="submit-button"
```
Comment (NoSloppy) [inline style_editor.html:10263]: Green to make it more obvious it's the "GO" button.

File: style_editor.html (hunk 2 above)
Comment (profezzorn) [reply style_editor.html:503]: nothing wrong with perfection, but putting everything in one PR is not perfection.

File: style_editor.html [hunk 9]
```diff
@@ -9280,37 +9494,220 @@
...
-function ClickInhilt() {
-  STATE_NUM_LEDS = 145 - STATE_NUM_LEDS;
-  var button = FIND("INHILT_BUTTON");
-  button.style.backgroundColor = STATE_NUM_LEDS == 1 ? "dodgerblue" : "";
-}
+/* Settings buttons saved as local cookies */
+var darkOn = true;
+var tipsOn = true;
+var stillOn = false;
+var swingEmOn = true;
```
Comment (profezzorn) [inline style_editor.html:9517]: This variable needs a better name.  (Because I don't get it, what is swingEm?)

File: style_editor.html [hunk 10]
```diff
@@ -8,7 +8,7 @@
   <title>ProffieOS Style Editor</title>
 <style>
 
-/* Make page elements non-selectable with drag box */
+/* Make page elements like buttons and whitespace non-selectable with drag box */
```
Comment (profezzorn) [inline style_editor.html:11]: This PR has WAY too much stuff in it.
Can we create a settings dialog with just *one* setting in it first?
Also, all the "minor cleanup" stuff is getting in the way of reviewing the settings code.

File: style_editor.html [hunk 11]
```diff
@@ -22,6 +22,110 @@
 * {
   box-sizing: border-box;
 }
+
+/* Dark mode styles */
```
Comment (profezzorn) [inline style_editor.html:26]: Dark mode should be a separate PR.

File: style_editor.html (hunk 1 above)
Comment (profezzorn) [reply style_editor.html:490]: universalizing should be a separate PR

File: style_editor.html [hunk 12]
```diff
@@ -9280,37 +9494,220 @@
...
+function onPageLoad() {
+  // Retrieve saved cookies and set buttons and settings accordingly
+  var body = document.querySelector("body");
+
+  // DARK_BUTTON
+  var darkButton = FIND("DARK_BUTTON");
+  var darkState = getSavedCookie("darkState");
+  var structuredView = FIND("structured_view")
+  // console.log("BC-getSavedCookie called, darkState = " + darkState);
+  if (darkState !== "false") {
```
Comment (profezzorn) [inline style_editor.html:9549]: There is a lot of duplication between onPageLoad() and ClickDark() (and other click functions)
Seems like we should have something like an SetDarkmode(onoff) function which can be called from both places.

File: style_editor.html (hunk 7 above)
Comment (profezzorn) [reply style_editor.html:9639]: I don't think we should do this.
Does it work on mobile?
Does it work on all browsers?
Does it work with screen readers?
Making things better is hard, it's often best to just leave them the way they are.

File: style_editor.html [hunk 13]
```diff
@@ -9323,6 +9720,11 @@
   if (v < 0) v += num_alternatives;
   if (v > num_alternatives) v -= num_alternatives;
   FIND("ALT").value = v;
+  // console.log("BC-Alt value changed: " + v);
+  if (!isNaN(v)) {
+    // console.log("BC-Saving cookie altValue: " + v);
+    saveCookie("altValue", v);
```
Comment (profezzorn) [inline style_editor.html:9726]: Hmm, saving the alt value seems weird to me.
It's not a setting really.

File: style_editor.html [hunk 14]
```diff
@@ -9340,6 +9742,15 @@
   FIND("VARIANT_VALUE").value = newValue;
   FIND("VARIANT_SLIDER").value = newValue;
   console.log("Updated Variant: " + newValue);
+  clearTimeout(variantTimeoutId);
+  // console.log("BC-Variant value changed: " + newValue);
+  // Variant cookie save delayed so rapid change doesn't do a bunch of saves
+  if (!isNaN(newValue)) {
+    variantTimeoutId = setTimeout(function() {
+      // console.log("BC-Saving cookie variantValue: " + newValue);
+      saveCookie("variantValue", newValue);
```
Comment (profezzorn) [inline style_editor.html:9751]: I don't think we should save the variant.
The variant kind of goes with the preset, so if you change the style, then I wouldn't expect to keep the variant.

**Comment by NoSloppy** [comment]
breaking into manageable pieces

File: style_editor.html (hunk 7 above)
Comment (NoSloppy) [reply style_editor.html:9639]: Yes.
Yes.
I don't have a screen reader.... can you build blade style's on that?
It's a safeguard in place. Example: If a user sticks WavLen as an AlphaL function, having no value creates a repeating loop of "get color" errors that will likely anger the browser.
It is working great, I was just saying I know it looks like a lot for a little value entry, but I threw the kitchen sink at it to make sure it's good.
